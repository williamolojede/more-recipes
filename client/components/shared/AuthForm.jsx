import React from 'react';
import PropTypes from 'prop-types';

/**
 * @class AuthForm
 * @extends {Component}
 */
class AuthForm extends React.Component {
  /**
   * @param {node} e - element
   * @return {view} - renders a form
   */
  onFormSubmit(e) {
    e.preventDefault();

    const email = this.emailInput.value;
    const password = this.passwordInput.value;
    if (this.props.type === 'login') {
      this.props.authFormSubmit({ email, password });
    }

    this.authForm.reset();
  }

  /**
   * @return {view} - renders a form
   */
  render() {
    return (
      <form
        className="auth-form form"
        ref={(node) => { this.authForm = node; }}
        onSubmit={e => this.onFormSubmit(e)}
      >
        {
          this.props.type === 'signup' ?
            <input
              type="text"
              name="usr-name"
              className="form__input"
              placeholder="name"
              required
            />
            :
            null
        }
        <input
          id="usr-email"
          type="email"
          name="usr-email"
          className="form__input"
          placeholder="email"
          required
          ref={(node) => { this.emailInput = node; }}
        />
        <input
          id="usr-pswd"
          type="password"
          name="usr-pswd"
          className="form__input"
          placeholder="password"
          required
          ref={(node) => { this.passwordInput = node; }}
        />

        <button type="submit" className="form__submit z-depth-1">
          {
            this.props.type === 'signup' ?
              'Register'
              :
              'Login'
          }
        </button>
      </form>
    );
  }
}

AuthForm.propTypes = {
  type: PropTypes.string.isRequired,
  authFormSubmit: PropTypes.func.isRequired,
};

export default AuthForm;
