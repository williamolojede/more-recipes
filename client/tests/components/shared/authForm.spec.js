import React from 'react';
import AuthForm from '../../../components/shared/AuthForm';

function setup() {
  const props = {
    type: 'login',
    authFormSubmit: jest.fn()
  };
  const wrapper = mount(<AuthForm {...props} />);

  return {
    wrapper
  };
}

describe('Nav Dropdown', () => {
  it('should render with right amount of elements for a login form', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with right amount of elements for a signup form', () => {
    const { wrapper } = setup();
    wrapper.setProps({ type: 'signup' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle login submit', () => {
    const { wrapper } = setup();
    const form = wrapper.find('form');
    const onFormSubmitSpy = jest.spyOn(wrapper.instance(), 'onFormSubmit');
    const authFormSubmitSpy = jest.spyOn(wrapper.instance().props, 'authFormSubmit');
    expect(form.length).toBe(1);
    form.simulate('submit', eventObject);
    expect(onFormSubmitSpy).toHaveBeenCalledTimes(1);
    expect(authFormSubmitSpy).toHaveBeenCalledWith({ email: '', password: '' });
  });

  it('should handle signup submit', () => {
    const { wrapper } = setup();
    wrapper.setProps({ type: 'signup' });
    const form = wrapper.find('form');
    const onFormSubmitSpy = jest.spyOn(wrapper.instance(), 'onFormSubmit');
    const authFormSubmitSpy = jest.spyOn(wrapper.instance().props, 'authFormSubmit');
    expect(form.length).toBe(1);
    form.simulate('submit', eventObject);
    expect(onFormSubmitSpy).toHaveBeenCalledTimes(1);
    expect(authFormSubmitSpy).toHaveBeenCalledWith({ email: '', password: '', fullname: '' });
  });
});
