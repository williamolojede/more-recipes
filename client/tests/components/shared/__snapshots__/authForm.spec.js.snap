// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Dropdown should render with right amount of elements for a login form 1`] = `
<AuthForm
  authFormSubmit={[Function]}
  type="login"
>
  <form
    className="auth-form form"
    onSubmit={[Function]}
  >
    <input
      className="form__input"
      id="usr-email"
      name="usr-email"
      placeholder="email"
      required={true}
      type="email"
    />
    <input
      className="form__input"
      id="usr-pswd"
      name="usr-pswd"
      placeholder="password"
      required={true}
      type="password"
    />
    <button
      className="form__submit z-depth-1"
      type="submit"
    >
      Login
    </button>
  </form>
</AuthForm>
`;

exports[`Nav Dropdown should render with right amount of elements for a signup form 1`] = `
<AuthForm
  authFormSubmit={[Function]}
  type="signup"
>
  <form
    className="auth-form form"
    onSubmit={[Function]}
  >
    <input
      className="form__input"
      name="usr-name"
      placeholder="fullname"
      required={true}
      type="text"
    />
    <input
      className="form__input"
      id="usr-email"
      name="usr-email"
      placeholder="email"
      required={true}
      type="email"
    />
    <input
      className="form__input"
      id="usr-pswd"
      name="usr-pswd"
      placeholder="password"
      required={true}
      type="password"
    />
    <button
      className="form__submit z-depth-1"
      type="submit"
    >
      Register
    </button>
  </form>
</AuthForm>
`;
